### PR TITLE
Jai/TLI 101.g -- QR Code Functionality to collaborate and show detailed graph and CSS

### LINKED ISSUE #103: TLI 101.g -- Create QR Code to Show Graph and Detailed Report
## **Description:**  
Develop a feature that allows users to collaborate on the dashboard by scanning a QR code. This functionality will enable users to view graphs and their detailed description collaboratively. The backend should be designed with clean architecture principles and include robust API support with proper error handling.

---

### **Frontend**  
- Design the QR code scanning interface.  
- Integrate QR code generation for sharing collaboration links.  
- Display graphs and data in a user-friendly format.  
- Provide feedback for successful or failed scans.  

### **Backend**  
- Design APIs for QR code generation and scanning.  
- Implement endpoints for retrieving and sharing dashboard data.  
- Ensure secure collaboration using token-based authentication or similar mechanisms.  
- Handle errors gracefully (e.g., invalid QR code, unauthorized access).  

### **Testing**  
- Test QR code scanning functionality on various devices.  
- Verify secure sharing and viewing of graphs.  
- Ensure error handling is consistent across edge cases.  

### **Documentation**  
- Write clear documentation for API endpoints and frontend integration.  
- Provide a user guide on generating and scanning QR codes for collaboration.  

---

### **QR Code Generation and Scanning**  
- Users can generate QR codes to share dashboard collaboration links.  
- QR codes can be scanned by other users to access the shared dashboard.  

### **Data Access and Display**  
- Collaborators can view the shared graphs in real-time.  
- Graphs via the QR code collaboration are displayed correctly.  

### **Security and Authentication**  
- Only authenticated users can generate or scan QR codes.  
- Shared links are secured using tokens, and data access is controlled.  

### **Error Handling**  
- Invalid or expired QR codes result in appropriate error messages.  
- Unauthorized access is prevented with proper error feedback.  

### **Usability**  
- The interface for generating and scanning QR codes is intuitive and responsive.  
- Collaboration setup is completed within a reasonable time frame (<10 seconds on average).

### PR DESCRIPTION
**Pull request by Jai0212**
This PR introduces a collaborative feature to the dashboard that allows users to scan QR codes to access shared graphs with detailed analysis. Upon scanning the QR code, users are redirected to a new page where they can view graphs, their descriptions, and a detailed report on them collaboratively.

### Key Features:
- **QR Code Generation**: Users can generate QR codes linked to collaboration URLs, which can be shared with others.
- **QR Code Scanning**: A clean, user-friendly interface is added to scan QR codes and display shared graphs and data.
- **Graph Display**: Once the QR code is scanned, users are directed to a page where the graph is displayed with detailed analysis.
- **CSS Styling**: Improved the interface's look and feel with custom CSS for a better user experience.
- **Clean Architecture**: Ensured Clean Architecture is maintained throughout this whole process
- **Mobile Optimization**: Optimized the CSS so that it looks good on phones as well

### Backend:
- APIs were designed for QR code generation and scanning.
- Endpoints handle retrieving and sharing dashboard data, with robust error handling for invalid or unauthorized QR codes.
- Backend functionality integrated using SOLID principles to follow Clean Architecture

### **SOLID Principles Used**

- **Single Responsibility Principle (SRP)**  
SRP was followed as the share functionality was added to the `UserRepositoryInterface`, which is responsible for user-related tasks such as login, share, and sign-up. By including the share functionality within the `UserRepositoryInterface`, the class maintains a single responsibility focused on user management.

- **Open/Closed Principle (OCP)**  
OCP was applied as the new share functionality was added by introducing a new function within the existing `UserRepositoryInterface`. This approach allows for extending functionality without modifying the existing codebase, making it easier to add new features in the future.

- **Dependency Inversion Principle (DIP)**  
DIP was implemented by adding a new function to the `UserRepositoryInterface` within the `use_cases` layer. This function was then implemented in the repositories layer, where the actual logic for the share functionality resides. The use case class calls this function, ensuring that high-level modules (use cases) depend on abstractions (interfaces), while low-level modules (repositories) implement these abstractions, maintaining the appropriate dependency flow.

- **Interface Segregation Principle (ISP)**  
ISP is followed by using smaller, specialized interfaces that include only the required functionality. This prevents any class or client from being forced to implement methods that are irrelevant to its purpose. The interfaces are kept focused, ensuring a clean and modular design.

### **Clean Architecture**  
Clean Architecture principles are followed, where the code is divided into distinct layers: interfaces, repositories, use cases, entities, and controllers. Each layer has a clear responsibility, with dependencies flowing in the correct direction. This separation ensures that:

- Each component is independent, making the codebase more modular and maintainable.
- High-level modules depend on abstractions, while low-level modules depend on the details of the implementation.
- The code remains flexible and easy to extend, allowing for changes without breaking existing functionality.

### **Design Patterns Used**

- **Facade Pattern**  
When adding the new QR Code functionality, the use of the Facade Pattern helped by ensuring that the repository functions related to user data were split into smaller, more focused interfaces such as `UserRepositoryInterface`. This allowed the new QR Code functionality to be added without interfering with other parts of the code. The `UserRepositoryInterface` acted as a facade, simplifying the complex repository logic into a clean, easy-to-use interface. This made it easier to extend the functionality by adding a new method for handling QR code generation, which could be implemented in the repository layer without disrupting the existing logic.

- **Dependency Injection (DI)**  
Dependency Injection was essential in adding the QR Code functionality because the use case for generating QR codes depended on the `UserRepositoryInterface` rather than a concrete implementation of the repository. By injecting the necessary repository functions through interfaces, I was able to add the new share functionality related to QR codes by simply extending the existing interface with a new method. The actual implementation of the new method was handled in the repository layer. This decoupled the high-level use cases from the low-level repository logic, making it easier to add and test the new functionality without modifying the existing codebase. DI also ensured that testing the new QR code generation functionality was straightforward, as I could inject mock repository implementations to isolate the functionality for unit testing.

### Acceptance Criteria:
- **QR Code Generation and Scanning**: Users can generate QR codes for collaboration links and others can scan them to access the shared dashboard.
- **Data Access and Display**: Collaborators can view graphs with a detailed analysis.
- **Security and Authentication**: Only authenticated users can generate or scan QR codes. Shared links are secured with tokens to control data access.
- **Error Handling**: Invalid or expired QR codes trigger proper error messages. Unauthorized access is prevented with clear feedback.
- **Usability**: The QR code scanning interface is intuitive and responsive, with collaboration setup completed in under 10 seconds on average.
- **Visuals**: The page looks visually good with proper styling on phones as well

### Additional Notes:
- This feature improves the collaborative functionality of the dashboard, enabling users to work together and analyze graphs effectively.

### Extra Feature
- Optimized frontend and backend to make the graph handle generate function much faster

image

image

### CODE REVIEW

File: frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            <button
+                className="share-button"
+                onClick={handleShare}
+                tabIndex={12}>
+                <svg
+                    xmlns="http://www.w3.org/2000/svg"
+                    viewBox="0 0 24 24"
+                    fill="currentColor"
+                    style={{ width: '16px', height: '16px', marginRight: '8px' }}>
+                    <path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7a3.248 3.248 0 000-1.39l7.11-4.11c.53.5 1.21.8 1.96.8 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .16.03.31.05.47L8.03 9.65c-.54-.51-1.25-.85-2.03-.85-1.66 0-3 1.34-3 3s1.34 3 3 3c.78 0 1.49-.34 2.03-.85l6.08 3.51c-.02.16-.05.31-.05.47 0 1.66 1.34 3 3 3s3-1.34 3-3-1.34-3-3-3z" />
```
Comment (Navnoorsinghmahal) [inline frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx:75]: What is the functionality of this part of the code?

File: backend/app/controllers/app.py [hunk 2]
```diff
@@ -534,5 +536,25 @@ def change_password():
         return jsonify({"code": 2, "error": True, "message": str(e)}), 500
 
 
+@app.route("/share/<encoded_data>", methods=["GET"])
```
Comment (Navnoorsinghmahal) [inline backend/app/controllers/app.py:539]: Is clean architecture followed while writing this code and which SOLID Principle is used here?

**Review by Navnoorsinghmahal — COMMENTED** [review]
Code Reviewed and looks great!!

File: backend/app/controllers/app.py (hunk 2 above)
Comment (Jai0212) [reply backend/app/controllers/app.py:539]: This PR adheres to **clean architecture** principles and implements **SOLID design principles**, ensuring a scalable, maintainable, and user-friendly solution for collaborative functionality.

---

## **SOLID Principles Implemented**

### 1. **Single Responsibility Principle (SRP)**
- **Frontend**:
  - **QR Code Generator Component**: Focuses solely on generating QR codes.
  - **QR Code Scanner Component**: Handles scanning QR codes and redirecting users to the appropriate page.
  - **Graph Display Page**: Responsible for rendering graphs, their descriptions, and detailed analysis.

---

### 2. **Open/Closed Principle (OCP)**
- Components are reusable and extendable:
  - The QR Code Generator and Scanner components are designed to handle future extensions, such as additional data formats or enhanced QR code features, without modifying existing code.
  - The graph display functionality can accommodate new types of graphs or analysis with minimal changes.
- There is no hardcoded data; all data is dynamically fetched from secure backend APIs.

---

### 3. **Liskov Substitution Principle (LSP)**
- All components can be replaced or extended without breaking the existing functionality:
  - The QR Code Scanner can be replaced with an alternative scanning library or module without affecting how data is handled.
  - The Graph Display Page ensures that any graph type adhering to the expected data structure can be displayed seamlessly.

---

## **Conclusion**
This feature enhances the collaborative functionality of the dashboard, enabling users to efficiently share and analyze graphs together. It maintains clean architecture and follows SOLID principles to ensure scalability and maintainability.

File: frontend/src/pages/SharePage/SharePage.jsx
Comment (tvirat) [inline frontend/src/pages/SharePage/SharePage.jsx:1]: Could you confirm if this code is associated with the page displayed to users after scanning the QR code?

File: frontend/src/pages/SharePage/SharePage.css [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+/* Responsive Design */
+@media (max-width: 768px) {
+
+    /* Data Container */
+    .data-container {
+        flex-direction: column;
+        /* Stack items vertically */
+        gap: 10px;
```
Comment (tvirat) [inline frontend/src/pages/SharePage/SharePage.css:179]: It's great that you implemented the media queries for small screen sizes; without them, the page's appearance would have been significantly impacted.

**Review by tvirat — CHANGES_REQUESTED** [review]

File: frontend/src/pages/SharePage/SharePage.jsx
Comment (Jai0212) [reply frontend/src/pages/SharePage/SharePage.jsx:1]: If you see the page QRCodeShare, when the user scans the QR Code or clicks on the hyperlink, he is navigated to a "/share/:encodedData" page. This is done using react-router-dom and can be seen in the App.jsx file. Once the user is taken to this link, the SharePage gets activated.

File: backend/demo2.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import base64
+import json
+
+# Data you want to share
+data = {
```
Comment (armaganngul) [inline backend/demo2.py:5]: are these demos gonna stay in the repo?

File: frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
```
Comment (armaganngul) [inline frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx:76]: Is this path the user's local machine or smth?

File: backend/app/repositories/interfaces.py [hunk 6]
```diff
@@ -93,3 +93,19 @@ def get_user_by_email_and_password(
     def update_password(self, email: str, new_password: str) -> None:
         """Updates the password of a user by their email."""
         pass
+
+    @abstractmethod
+    def process_shared_data(self, encoded_data: str) -> dict:
```
Comment (akshatoza1) [inline backend/app/repositories/interfaces.py:98]: What's this change for specifically? Like the interface of what?

File: frontend/src/pages/SharePage/SharePage.jsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        // Sum all the accuracy values
+        const totalAccuracy = data.reduce((sum, item) => sum + item.accuracy, 0);
+
+        // Calculate the average
+        const averageAccuracy = totalAccuracy / data.length;
+
+        setAverageBias(parseFloat(averageAccuracy.toFixed(2)));
+    };
+
+    const maxValue = () => {
```
Comment (armaganngul) [inline frontend/src/pages/SharePage/SharePage.jsx:109]: what's the purpose of this?

**Review by armaganngul — COMMENTED** [review]
Looks good Jai! were you able to figure out the dimension issue in mobile devices?

File: frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.css [hunk 8]
```diff
@@ -0,0 +1,27 @@
+.share-button {
```
Comment (akshatoza1) [inline frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.css:1]: Anything specific to be noted about this CSS for the QR code?

File: backend/demo2.py (hunk 4 above)
Comment (Jai0212) [reply backend/demo2.py:5]: No, this is only for testing purposes. Before our final publish, we will delete the demo files

File: frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx (hunk 5 above)
Comment (Jai0212) [reply frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.jsx:76]: This is an SVG for the share icon so that we can make the share button look good.

File: frontend/src/pages/SharePage/SharePage.jsx (hunk 7 above)
Comment (Jai0212) [reply frontend/src/pages/SharePage/SharePage.jsx:109]: This is for finding the maximum bias in the graph as this needs to be passed onto the chart component to decide which bar should be highlighted in red and which shouldn't. It is an indication to the engineers if their bias threshold is crossed or not. This value is being used by the slider component as well.

**Comment by Jai0212** [comment]
> Looks good Jai! were you able to figure out the dimension issue in mobile devices?

Ya, I tested out various mobile device sizes and wrote specific media queries for each mobile size in order to optimize the visuals.

File: backend/app/repositories/interfaces.py (hunk 6 above)
Comment (Jai0212) [reply backend/app/repositories/interfaces.py:98]: Since we are adding a new use case, we've adopted **Clean Architecture** to ensure proper separation of concerns. To achieve this:

1. I added an abstract method to the interface, which is implemented in the `UserRepository`.  
   - This follows the **Dependency Inversion Principle (DIP)**, where high-level modules (use cases) depend on abstractions (interfaces) instead of concrete implementations (repositories).
   
2. The use case calls this method through the repository interface.  
   - This follows the **Open/Closed Principle (OCP)**, where the system is open for extension (adding new use cases) but closed for modification (existing components remain unchanged).
   
3. Finally, in `app.py` (the controller), the appropriate use case is instantiated and called, ensuring a clean flow of data and logic across the application layers.  
   - This adheres to the **Single Responsibility Principle (SRP)**, ensuring that each class or module has one reason to change and is responsible for a single aspect of the functionality.

By following **Clean Architecture** and applying **SOLID principles**, we ensure that the system remains modular, maintainable, and scalable.

File: frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.css (hunk 8 above)
Comment (Jai0212) [reply frontend/src/pages/DashboardPage/QRCodeShare/QRCodeShare.css:1]: The share button isn't using an image, rather, it is using an SVG for better clarity. THe rest of the CSS properties are quite general and are created to maintain the existing theme of the page.

File: frontend/src/pages/SharePage/SharePage.jsx
Comment (tvirat) [reply frontend/src/pages/SharePage/SharePage.jsx:1]: Okay, makes sense!
